IdentityCore/src/cache/mac/MSIDMacKeychainTokenCache.m

```diff
@@ -516,17 +567,138 @@ - (MSIDCredentialCacheItem *)tokenWithKey:(__unused MSIDCacheKey *)key
+    
+    if (!key)
+    {
+        [self createError:@"Key is nil."
+                   domain:MSIDErrorDomain errorCode:MSIDErrorInvalidDeveloperParameter error:error context:context];
+        return NO;
+    }
+    
+    NSMutableDictionary *query = nil;
+    
+    if (key.isShared)
+    {
+        query = [self sharedItemPrimaryAttributesForKey:key];
+    }
+    else
+    {
+        query = [self secretItemPrimaryAttributesForKey:key];
+    }
+    
+    [query addEntriesFromDictionary:[self secondaryAttributesForKey:key]];
+    query[(id)kSecMatchLimit] = (id)kSecMatchLimitAll;
+    
+    MSID_LOG_INFO(context, @"Trying to delete keychain items...");
+    OSStatus status = SecItemDelete((CFDictionaryRef)query);
+    MSID_LOG_INFO(context, @"Keychain delete status: %d", (int)status);
+    
+    if (status != errSecSuccess && status != errSecItemNotFound)
+    {
+        [self createError:@"Failed to remove multiple accounts from keychain"
+                   domain:MSIDKeychainErrorDomain errorCode:status error:error context:context];
+        return NO;
+    }
+    
+    return YES;
+}
+
+- (NSArray *)itemsWithKey:(MSIDCacheKey *)key
+                  context:(id<MSIDRequestContext>)context
+                    error:(NSError **)error
+{
+    MSID_TRACE;
+    MSID_LOG_VERBOSE(context, @"Get keychain items, key info (account: %@ service: %@ generic: %@ type: %@, keychainGroup: %@)", _PII_NULLIFY(key.account), key.service, _PII_NULLIFY(key.generic), key.type, [self keychainGroupLoggingName]);
+    MSID_LOG_VERBOSE_PII(context, @"Get keychain items, key info (account: %@ service: %@ generic: %@ type: %@, keychainGroup: %@)", key.account, key.service, key.generic, key.type, self.keychainGroup);
+    
+    NSMutableDictionary *cacheQuery = nil;
+    
+    if (key.isShared)
+    {
+        cacheQuery = [self sharedItemPrimaryAttributesForKey:key];
+    }
+    else
+    {
+        cacheQuery = [self secretItemPrimaryAttributesForKey:key];
+    }
+    
+    [cacheQuery addEntriesFromDictionary:[self secondaryAttributesForKey:key]];
+    
+    // Per Apple's docs, kSecReturnData can't be combined with kSecMatchLimitAll:
+    // https://developer.apple.com/documentation/security/1398306-secitemcopymatching?language=objc
+    // For this reason, we retrieve references to the items, then (below) use a second SecItemCopyMatching()
+    // to retrieve the data for each, deserializing each into an account object.
+    NSMutableDictionary *query = [cacheQuery mutableCopy];
+    query[(id)kSecMatchLimit] = (id)kSecMatchLimitAll;
+    query[(id)kSecReturnRef] = @YES;
+    
+    CFTypeRef cfItems = nil;
+    MSID_LOG_INFO(context, @"Trying to find keychain items...");
+    OSStatus status = SecItemCopyMatching((CFDictionaryRef)query, &cfItems);
+    MSID_LOG_INFO(context, @"Keychain find status: %d", (int)status);
+    
+    if (status == errSecItemNotFound)
+    {
+        MSID_LOG_INFO(context, @"Account not found");
```

Comment:
why does it say "account not found"? It can be a token just as well

Reply:
I still need to fix the log statements . The PR is just an initial commit to see how we can differentiate between different kinds of credentials.